pmlc/dialect/abi/ir/loop_op.cc

```diff
@@ -0,0 +1,184 @@
+    unsigned idx = 0;
+    for (auto tyAttr : networkFieldTypes) {
+      auto val = networkOp.getOperand(idx);
+      auto definingOp = val.getDefiningOp();
+      if (definingOp && definingOp->hasTrait<mlir::OpTrait::ConstantLike>()) {
+        // N.B. Constant-Like operations have no side-effects, exactly one
+        //      result, and no operands.
+        auto *denormOp = definingOp->clone();
+        rewriter.insert(denormOp);
+        auto arg = loopOp.getBodyEntryBlock()->getArgument(idx);
+        arg.replaceAllUsesWith(denormOp->getResult(0));
+        loopOp.getBodyEntryBlock()->eraseArgument(idx);
+        loopOp.getFiniEntryBlock()->eraseArgument(idx);
+        networkOp.getOperation()->eraseOperand(idx);
+      } else {
+        newNetworkFieldTypes.emplace_back(tyAttr);
+        ++idx;
+      }
+    }
+    if (networkFieldTypes.size() != newNetworkFieldTypes.size()) {
+      loopOp.setAttr(abi::LoopOp::getNetworkFieldTypesAttrName(),
+                     rewriter.getArrayAttr(newNetworkFieldTypes));
+      rewriter.finalizeRootUpdate(loopOp);
+      return mlir::success();
+    }
+    rewriter.cancelRootUpdate(loopOp);
+    return mlir::failure();
+  }
+};
+
+struct RemoveUnusedNetworkFieldsPattern final
+    : public mlir::OpRewritePattern<LoopOp> {
+  using mlir::OpRewritePattern<LoopOp>::OpRewritePattern;
+
+  mlir::LogicalResult match(LoopOp loopOp) const final {
+    auto networkOp = loopOp.getInitTerminator();
+    auto *bodyEntryBlock = loopOp.getBodyEntryBlock();
+    auto *finiEntryBlock = loopOp.getFiniEntryBlock();
+    for (unsigned idx = 0; idx < networkOp.getNumOperands(); ++idx) {
+      if (bodyEntryBlock->getArgument(idx).use_empty() &&
+          finiEntryBlock->getArgument(idx).use_empty()) {
+        return mlir::success();
+      }
+    }
+    return mlir::failure();
+  }
+
+  void rewrite(LoopOp loopOp, mlir::PatternRewriter &rewriter) const final {
+    rewriter.updateRootInPlace(loopOp, [&] {
+      auto networkOp = loopOp.getInitTerminator();
+      auto *bodyEntryBlock = loopOp.getBodyEntryBlock();
+      auto *finiEntryBlock = loopOp.getFiniEntryBlock();
+      unsigned idx = 0;
+      while (idx < networkOp.getNumOperands()) {
+        if (bodyEntryBlock->getArgument(idx).use_empty() &&
+            finiEntryBlock->getArgument(idx).use_empty()) {
+          networkOp.getOperation()->eraseOperand(idx);
+          bodyEntryBlock->eraseArgument(idx);
+          finiEntryBlock->eraseArgument(idx);
+        } else {
+          ++idx;
+        }
+      }
+      loopOp.setNetworkFieldTypes(networkOp.getOperandTypes());
+    });
+  }
+};
+
+} // namespace
+
+void LoopOp::getCanonicalizationPatterns(
+    mlir::OwningRewritePatternList &patterns, mlir::MLIRContext *ctx) {
+  patterns.insert<DenormalizeConstantsPattern>(ctx);
+  patterns.insert<RemoveUnusedNetworkFieldsPattern>(ctx);
+}
+
+mlir::Region &LoopOp::getLoopBody() { return bodyRegion(); }
+
+bool LoopOp::isDefinedOutsideOfLoop(mlir::Value value) {
```

Comment:
You can probably use `mlir/Transforms/RegionUtils.h`, specifically the `getUsedValuesDefinedAbove` or `getUsedValuesDefinedAbove` helper functions.

Reply:
Awesome -- thanks!